### PR TITLE
Nonholonomic Variable-Mass Rocket Car Example

### PR DESCRIPTION
**Pull request by abhikamath**
Example
-------

   - [x] There are no merge conflicts with the `master` branch.
   - [x] Examples rst file is placed in `docs/examples/`.
   - [x] An entry with a 200x200 pixel image has been added to
     `docs/index.rst`.
   - [x] RsT and image filenames are kebab-case (lower case hyphen-separated).
   - [x] jupyter_sphinx RsT directives are used to ensure example is executed
     on build.
   - [x] jupyter_sphinx `ipynb` and `py` download links are included.
   - [x] The example has adequate text and figures (preferably SVG) explaining
     the problem.
   - [ ] All reviewer comments have been addressed.

### CODE REVIEW

File: docs/examples/Rocket_Car.rst [hunk 1]
```diff
@@ -0,0 +1,1585 @@
...
+
+.. raw:: html
+
+   <center>
+
+:math:`\mathbf{\text{Modeling of a Variable-Mass Nonholonomic Gyrostatic Rocket Car Using Extended Kane's Equations}}`
+
+.. raw:: html
+
+   </center>
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:11]: Headings need to be convereted to standard RST like the other examples. This should be:

```rst
============================================================================
Modeling of a Variable-Mass Nonholonomic Gyrostatic Rocket Car Using Extended Kane's Equations
=============================================================================
```

File: docs/examples/Rocket_Car.rst [hunk 2]
```diff
@@ -0,0 +1,1585 @@
...
+
+#####
+
+.. raw:: html
+
+   <center>
+
+:math:`\text{Ge, Z.M. and Cheng, Y.H., 1982. Extended Kane’s equations for nonholonomic variable mass system.}`
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:19]: Make a rst citation for this:

https://docutils.sourceforge.io/docs/user/rst/quickref.html#citations

File: docs/examples/Rocket_Car.rst (hunk 1 above)
Comment (moorepants) [reply docs/examples/Rocket_Car.rst:11]: Also, we shouldn't copy the title of the journal paper if that is what you've done here. Make your own title for this example and cite the other paper as the inspiration.

File: docs/examples/Rocket_Car.rst [hunk 3]
```diff
@@ -0,0 +1,1585 @@
...
+
+   </center>
+
+#####
+
+.. raw:: html
+
+   <center>
+
+:math:`\text{Figure 1: Idealized model of a jet racing car that is propelled by a rocket engine at point P, the rocket engine being treated as a variable mass particle at P}`
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:31]: Figures need to use the proper rst syntax:

https://docutils.sourceforge.io/docs/ref/rst/directives.html#figure

File: docs/examples/Rocket_Car.rst [hunk 4]
```diff
@@ -0,0 +1,1585 @@
...
+
+   </center>
+
+#####
+
+.. raw:: html
+
+   <center>
+
+:math:`\text{Leimanis, E., 2013. The general problem of the motion of coupled rigid bodies about a fixed point (Vol. 7). Springer Science & Business Media. (Page 207)}`
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:55]: Use a rst citation.

File: docs/examples/Rocket_Car.rst [hunk 5]
```diff
@@ -0,0 +1,1585 @@
...
+
+.. raw:: html
+
+   </center>
+
+.. code:: ipython3
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:61]: All code input should use the ``.. jupyter-execute::`` directive. See the other examples for usage: https://raw.githubusercontent.com/pydy/pydy/master/docs/examples/carvallo-whipple.rst

File: docs/examples/Rocket_Car.rst [hunk 6]
```diff
@@ -0,0 +1,1585 @@
...
+
+    Q.set_vel(N, Q.pos_from(O).dt(N))
+    Q.vel(N)
+
+
+
+
+.. math::
+
+    \displaystyle \dot{q}_{1}\mathbf{\hat{n}_x} + \dot{q}_{2}\mathbf{\hat{n}_y}
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:119]: No code output should be included. https://github.com/jupyter/jupyter-sphinx will create the outputs when sphinx is run to build the documentation website.

File: docs/examples/Rocket_Car.rst (hunk 1 above)
Comment (moorepants) [reply docs/examples/Rocket_Car.rst:11]: Use this heading level method to match the other files:

```rst
=============
Title of page
=============
Heading
=======

Subheading
----------

File: docs/examples/Rocket_Car.rst [hunk 7]
```diff
@@ -0,0 +1,1585 @@
...
+
+   <center>
+
+:math:`\text{Kane, T.R., 1978. Nonholonomic multibody systems containing gyrostats. In Dynamics of Multibody Systems (pp. 97-107). Springer, Berlin, Heidelberg.}`
+
+.. raw:: html
+
+   </center>
+
+:math:`\text{Naming Convention:}`
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:1023]: All of these are in math directives for some reason. All normal text needs to be extracted from these and use proper rst notation.

**Comment by moorepants** [comment]
- You've commited two Macintosh `.DS_Store` files which need to be removed.
- We can't include any images you've taken from other sources due to copyright. Remove any that are and replace with your own text or simply a citation to the relevant information. You maybe able to link to a Wikipedia article one gyrostats ore something if that is helpful.

These basic comments will apply to the other PR too. So, fix them both up so that they are well formatted RST with the jupyter-sphinx cells and then let me know and I'll review again.

File: docs/examples/Rocket_Car.rst [hunk 8]
```diff
@@ -0,0 +1,875 @@
...
+    sollist_values = list(solution.values())
+    sollist_values
+
+
+
+.. jupyter-execute::
+
+    sollist_values_simple = []
+    for i in range(8):
+       sollist_values_simple.append(sm.factor(TR2(sollist_values[i]).simplify()))
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:338]: This line is giving an error:

```
Extension error:

Cell raised uncaught exception:

---------------------------------------------------------------------------

IndexError                                Traceback (most recent call last)

<ipython-input-49-e76e181d41dd> in <module>()

      1 sollist_values_simple = []

      2 for i in range(8):

----> 3    sollist_values_simple.append(sm.factor(TR2(sollist_values[i]).simplify()))

IndexError: list index out of range
```

Maybe change `8` to `len(something)` so you have the correct range?

File: docs/examples/Rocket_Car.rst [hunk 9]
```diff
@@ -0,0 +1,875 @@
...
+
+
+.. jupyter-execute::
+
+    kane_2_eq = sm.Eq(kane_2.simplify().subs(soldict).simplify().subs(u).simplify(), 0)
+    # kane_2_eq # Invalid JSON
+
+
+References
+==========
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:872]: This gave this warning:

```
/home/travis/build/pydy/pydy/pydy/codegen/ode_function_generators.py:docstring of pydy.codegen.ode_function_generators.generate_ode_function:97: WARNING: Block quote ends without a blank line; unexpected unindent.

/home/travis/build/pydy/pydy/docs/examples/Rocket_Car.rst:871: WARNING: Title level inconsistent:

References

==========
```

**Comment by moorepants** [comment]
Don't forget:

> You've commited two Macintosh .DS_Store files which need to be removed.

**Comment by moorepants** [comment]
You are getting this error:

```
Extension error:

Cell raised uncaught exception:

---------------------------------------------------------------------------

KeyError                                  Traceback (most recent call last)

<ipython-input-89-6419ec7993f0> in <module>

      2 C71, C81 = sm.symbols('C_71, C_81')

      3 C_71 = soldict[q7.diff()].diff(u1)

----> 4 C_81 = soldict[q8.diff()].diff(u1)

      5 Fstar_1_G2_I = -J2 * (N_w_A2.dot(q7.diff().diff() * A2.y + q7.diff()*(N_w_A2.dot(A2.x)*A2.z - N_w_A2.dot(A2.z)*A2.x)) + C_71 * (N_w_A2.dot(A2.y).diff() + q7.diff().diff())) \

      6                -J2 * (N_w_A2.dot(q8.diff().diff() * A2.y + q8.diff()*(N_w_A2.dot(A2.x)*A2.z - N_w_A2.dot(A2.z)*A2.x)) + C_81 * (N_w_A2.dot(A2.y).diff() + q8.diff().diff()))   # B1 \ B2
```

q8.diff() doesn't seem to be present in `solidict`.

File: docs/examples/Rocket_Car.rst [hunk 10]
```diff
@@ -0,0 +1,874 @@
+==============================================================================================
+Modeling of a Variable-Mass Nonholonomic Gyrostatic Rocket Car Using Extended Kane’s Equations
+==============================================================================================
+
+This is an implementation of the nonholonomic rocket-engine-powered jet racing car example from [Ge1982]_. This example provides insight into analytically modeling variable-mass systems using Kane's method and the extended Kane's equations for variable-mass systems. It also demonstrates the efficacy of Kane's method in the modeling of complex dynamical systems.
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:5]: wrap this to 79 character line length

File: docs/examples/Rocket_Car.rst [hunk 11]
```diff
@@ -0,0 +1,874 @@
+==============================================================================================
+Modeling of a Variable-Mass Nonholonomic Gyrostatic Rocket Car Using Extended Kane’s Equations
+==============================================================================================
+
```
Comment (moorepants) [inline docs/examples/Rocket_Car.rst:4]: Add the download links at the top like the other examples.

**Comment by moorepants** [comment]
Looks like there is an error in the code:

```
ell raised uncaught exception:

---------------------------------------------------------------------------

KeyError                                  Traceback (most recent call last)

<ipython-input-89-6419ec7993f0> in <module>

      2 C71, C81 = sm.symbols('C_71, C_81')

      3 C_71 = soldict[q7.diff()].diff(u1)

----> 4 C_81 = soldict[q8.diff()].diff(u1)

      5 Fstar_1_G2_I = -J2 * (N_w_A2.dot(q7.diff().diff() * A2.y + q7.diff()*(N_w_A2.dot(A2.x)*A2.z - N_w_A2.dot(A2.z)*A2.x)) + C_71 * (N_w_A2.dot(A2.y).diff() + q7.diff().diff())) \

      6                -J2 * (N_w_A2.dot(q8.diff().diff() * A2.y + q8.diff()*(N_w_A2.dot(A2.x)*A2.z - N_w_A2.dot(A2.z)*A2.x)) + C_81 * (N_w_A2.dot(A2.y).diff() + q8.diff().diff()))   # B1 \ B2

KeyError: Derivative(q8(t), t)
```

**Comment by moorepants** [comment]
I get the same error locally. Are you building the docs locally to see that? I then removed all the lines below and including the `C_81` and built the page and I see that q.diff() is not being solved for:
image
If you know those equations are linear in the qdots then maybe try solving them as a linear system Ax=b.

**Comment by moorepants** [comment]
You also have 10 equations there and only 8 unknowns.

**Comment by abhikamath** [comment]
Screen Shot 2021-03-07 at 6 10 52 PM

I am building the docs locally—it seems to work fine on my machine, for some reason. 

I'll look into it.

**Review by abhikamath — COMMENTED** [review]
That didn't work either.

**Comment by moorepants** [comment]
What version of sympy are you using? Maybe something has changed in newer sympy releases? I was using 1.7.

**Comment by abhikamath** [comment]
> What version of sympy are you using? Maybe something has changed in newer sympy releases? I was using 1.7.

I'm using version 1.5.

**Comment by moorepants** [comment]
This may be a bug that should be reported to sympy, but for now do something like this:

```python
eqs = Matrix([eq1, ..., eq8])
qdots = Matrix([q1d, ..., q8d])
A = eqs.jacobian(qdots)
b = - eqs.xreplace({qdi: 0 for qdi in qdots})
sol = {qdi: soli for qdi, soli in zip(qdots, A.LUsolve(b))}
```

**Comment by moorepants** [comment]
The appveyor error may be that pythreejs no longer support python 2. We can ignore that and I'll fix it elsewhere.

I don't think the table on the main page is correct.

File: docs/index.rst [hunk 12]
```diff
@@ -60,20 +60,31 @@ Table of Contents
              :target: examples/carvallo-whipple.html
 
              Carvallo-Whipple bicycle model.
-      * - .. figure:: examples/kane-levinson-1985.png
+      * - .. figure:: examples/astrobee.png
```
Comment (moorepants) [inline docs/index.rst:63]: This wasn't showing up in my local build, I don't think you have it defined correctly.

**Comment by moorepants** [comment]
Here is the error:
```
/home/moorepants/src/pydy/docs/index.rst:4: WARNING: Error parsing content block for the "list-table" directive: uniform two-level bullet list expected, but row 3 does not contain the same number of items as row 1 (3 vs 2).

.. list-table::

   * - .. figure:: examples/mass_spring_damper.svg
          :width: 200px
          :target: examples/mass-spring-damper.html

          Linear mass-spring-damper system with gravity.
     - .. figure:: examples/multidof-holonomic.png
          :width: 200px
          :target: examples/multidof-holonomic.html

          A double compound and simple pendulum.
   * - .. figure:: examples/three-link-conical-pendulum.gif
          :width: 200px
          :target: examples/three-link-conical-pendulum.html

          Three link conical compound pendulum.
     - .. figure:: examples/bicycle-geometry.png
          :width: 200px
          :target: examples/carvallo-whipple.html

          Carvallo-Whipple bicycle model.
   * - .. figure:: examples/astrobee.png
          :width: 200px
          :target: examples/astrobee.html

          Astrobee free-flying ISS robot.
     - .. figure:: examples/kane-levinson-1985.png
          :width: 200px
          :target: examples/kane-levinson-1985-chapter-02.html

          Nonholonomic Rocket Car.
     - .. figure:: examples/rocket-car.png
          :width: 200px
          :target: examples/rocket-car.html

          Exercises from Chapter 2 of Kane & Levinson 1985.
   * - .. figure:: examples/kane-levinson-1985.png
          :width: 200px
          :target: examples/kane-levinson-1985-chapter-03.html

          Exercises from Chapter 3 of Kane & Levinson 1985.
```

**Comment by moorepants** [comment]
I fixed the table.

**Comment by moorepants** [comment]
Some of the equations have too large width (mathjax doesn't seem to wrap them). Either figure out how to wrap them, don't display them, or use some method to make them print more nicely.

Here is a function I've used to display symbols equal to some longer expressions:

```python
from IPython.display import Latex, display
def print_answer(x, x_res):
    for xi, xri in zip(x, x_res):
        display(Latex(sm.latex(sm.Eq(xi, xri), mode='inline')))
```

That could be a solution for the dictionaries not printing.

**Comment by moorepants** [comment]
It passes now. Do you think you can make some adjustments for those very wide outputs?

**Comment by moorepants** [comment]
Ok, let's go with this! Thanks for all the hard work here. It's a really nice example.

**Comment by abhikamath** [comment]
I've pushed another commit that wraps the remaining overflowing equations as well. Additionally, the diagram (SVG) doesn't seem to be rendered properly (the arrows)—do you have any suggestions to fix that?

**Comment by moorepants** [comment]
https://pydy.readthedocs.io/en/latest/examples/rocket-car.html

You'll need to open a new PR to make any more fixes.

I see that the inertial frame arrows look odd. I'm not sure why they would render differently than other arrows if you create them in the same way. In inkscape in the "object properties" you can set different arrow heads. That's all I know.

**Pull request merged**